src/Windows/Markdown/Editor/Impl/Preview/Code/RCodeBlockRenderer.cs

```diff
@@ -0,0 +1,224 @@
+using System;
+using System.Collections.Generic;
+using System.Drawing.Imaging;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Markdig.Renderers;
+using Markdig.Renderers.Html;
+using Markdig.Syntax;
+using Microsoft.Common.Core;
+using Microsoft.Common.Core.Disposables;
+using Microsoft.Common.Core.Services;
+using Microsoft.Markdown.Editor.Settings;
+using Microsoft.R.Components.InteractiveWorkflow;
+using Microsoft.R.Components.Settings;
+using Microsoft.R.Host.Client;
+using mshtml;
+using Microsoft.Markdown.Editor.Preview.Parser;
+using static System.FormattableString;
+
+namespace Microsoft.Markdown.Editor.Preview.Code {
+    /// <summary>
+    /// Renders R code block output into HTML element
+    /// </summary>
+    internal sealed class RCodeBlockRenderer : HtmlObjectRenderer<CodeBlock>, IDisposable {
+        private static string _dynamicPlaceholderImage;
+
+        /// <summary>
+        /// Rendered results cache. Caches HTML for every code block.
+        /// Blocks are differentiated by their order of appearance
+        /// and the contents hash.
+        /// </summary>
+        private readonly List<RCodeBlock> _blocks = new List<RCodeBlock>();
+
+        private readonly IServiceContainer _services;
+        private readonly IRMarkdownEditorSettings _settings;
+        private readonly string _sessionId;
+        private readonly CancellationTokenSource _hostStartCts = new CancellationTokenSource();
+        private readonly RSessionCallback _sessionCallback = new RSessionCallback();
+
+        private CancellationTokenSource _blockEvalCts = new CancellationTokenSource();
+        private IRSession _session;
+        private int _blockNumber;
+        private Task _evalTask = Task.CompletedTask;
+
+        public RCodeBlockRenderer(string documentName, IServiceContainer services) {
+            _services = services;
+            _settings = services.GetService<IRMarkdownEditorSettings>();
+            _sessionId = Invariant($"({documentName} - {Guid.NewGuid()}");
+            StartSessionAsync(_hostStartCts.Token).DoNotWait();
+        }
+
+        public IDisposable StartRendering() {
+            _blockNumber = 0;
+            _blockEvalCts?.Cancel();
+            _blockEvalCts = new CancellationTokenSource();
+            return Disposable.Create(() => _evalTask = _settings.AutomaticSync ? EvaluateBlocksAsync(_blockEvalCts.Token) : Task.CompletedTask);
+        }
+
+        #region Evaluation
+        private Task EvaluateBlockAsync(IRSession session, int blockNumber, CancellationToken ct) {
+            var block = _blocks[blockNumber];
+            block.State = CodeBlockState.Created;
+            return block.EvaluateAsync(session, _sessionCallback, ct);
+        }
+
+
+        private Task EvaluateBlocksAsync(CancellationToken ct) {
+            var blocks = _blocks.ToArray();
+            //TODO: clear session on cache drop
+            return Task.Run(async () => {
```

Comment:
I don't think you need a `Task.Run`. `await TaskUtilities.SwitchToBackgroundThread()` should be enough.